File: scripts/check_plugin_version.py
#!/usr/bin/env python3
"""The version has to move when the plugin does.

On 11 August 2026 the map emitter was changed — a change to what a reading
CONTAINS — and shipped under 0.9.0, the version it already had. Claude Code
compares the version you have installed against the version on offer, saw the
same number, and correctly refused to update. The user's report was "I'm trying
to update the plugin version in Claude, but it does not seem to want to."

That is the second time. `6396814` records the first ("Bump to 0.8.0 — three
skills changed under a version that never moved"), and its message already said
why it matters: a version that does not move reports agreement that is not
there. Saying it in a commit message did not stop it happening again, so it is
a check now.

Two invariants, deliberately separate:

  MARKERS   Every SKILL.md's `vibeassist-skill-version` marker equals the
            version in plugin.json. Needs no git history — a bump that edits
            plugin.json and forgets a marker ships four skills claiming a
            version their own plugin does not.

  MOVED     If anything under plugins/ changed against the base branch, the
            version in plugin.json must differ from the base's. This is the one
            that catches the failure above.

MOVED needs a base to compare against, so it only runs when `--base` is given
(CI passes the pull request's base). MARKERS always runs. A local run with no
arguments therefore checks what it can and says what it skipped, rather than
passing silently and implying more than it looked at.

Exit 0 = every invariant it could check held. Exit 1 = a real disagreement.
Exit 2 = the check could not run (bad ref, unreadable manifest) — which is NOT
the same as passing, and CI must treat it as failure.
"""

from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
PLUGIN_DIR = Path("plugins/vibeassist")
MANIFEST = PLUGIN_DIR / ".claude-plugin" / "plugin.json"

# The marker each packaged SKILL.md carries, which the app's connected session
# compares against the version VA serves. Captured, so a mismatch can name both
# numbers rather than just saying "wrong".
MARKER = re.compile(r"<!--\s*vibeassist-skill-version:\s*(?P<version>[0-9][0-9.]*)")

# Files under the plugin that do NOT change what an installed copy does. A
# version bump for a typo in the README is noise, and a check that cries wolf
# gets bypassed — which costs more than the typo. Everything else counts.
IGNORED = {MANIFEST}


def fail(msg: str) -> None:
    print(f"FAIL  {msg}")


def run_git(args: list[str]) -> tuple[int, str]:
    p = subprocess.run(
        ["git", *args], cwd=REPO_ROOT, capture_output=True, text=True, check=False
    )
    return p.returncode, (p.stdout or "") + (p.stderr or "")


def version_of(text: str, where: str) -> str | None:
    try:
        v = json.loads(text).get("version")
    except json.JSONDecodeError as e:
        fail(f"{where} is not valid JSON: {e}")
        return None
    if not isinstance(v, str) or not v.strip():
        fail(f"{where} has no `version` string.")
        return None
    return v.strip()


def check_markers(version: str) -> bool:
    """Every SKILL.md marker equals the manifest version."""
    skills = sorted((REPO_ROOT / PLUGIN_DIR / "skills").glob("*/SKILL.md"))
    if not skills:
        fail(f"no SKILL.md found under {PLUGIN_DIR / 'skills'} — has the layout moved?")
        return False

    ok = True
    for path in skills:
        rel = path.relative_to(REPO_ROOT)
        m = MARKER.search(path.read_text(encoding="utf-8"))
        if not m:
            fail(
                f"{rel} carries no `vibeassist-skill-version` marker, so an installed "
                f"copy cannot tell the app which version it is running."
            )
            ok = False
            continue
        found = m.group("version")
        if found != version:
            fail(
                f"{rel} says {found}, but {MANIFEST} says {version}. "
                f"One of them is lying to whoever installed it."
            )
            ok = False
    if ok:
        print(f"ok    {len(skills)} SKILL.md marker(s) agree with {MANIFEST}: {version}")
    return ok


def check_moved(base: str, version: str) -> bool:
    """Anything changed under plugins/ ⇒ the version differs from the base's."""
    code, out = run_git(["rev-parse", "--verify", f"{base}^{{commit}}"])
    if code != 0:
        fail(f"cannot resolve base ref {base!r} — nothing to compare against.\n{out.strip()}")
        sys.exit(2)

    code, out = run_git(["diff", "--name-only", f"{base}...HEAD", "--", str(PLUGIN_DIR)])
    if code != 0:
        fail(f"could not diff against {base}.\n{out.strip()}")
        sys.exit(2)

    changed = [Path(line) for line in out.splitlines() if line.strip()]
    material = [p for p in changed if p not in IGNORED]
    if not material:
        print(f"ok    nothing under {PLUGIN_DIR} changed against {base} — no bump needed")
        return True

    code, base_manifest = run_git(["show", f"{base}:{MANIFEST}"])
    if code != 0:
        # The manifest not existing on the base is a new plugin, not a stalled
        # version. Nothing to compare, nothing to complain about.
        print(f"ok    {MANIFEST} is new against {base} — no previous version to move from")
        return True

    was = version_of(base_manifest, f"{MANIFEST} at {base}")
    if was is None:
        sys.exit(2)

    if was == version:
        listing = "\n".join(f"        {p}" for p in material)
        fail(
            f"{len(material)} file(s) under {PLUGIN_DIR} changed, but the version is still "
            f"{version}:\n{listing}\n"
            f"        Anyone already on {version} keeps a different {version} and "
            f"`/plugin update vibeassist` has nothing to move to.\n"
            f"        Bump `version` in {MANIFEST} and the marker in every SKILL.md."
        )
        return False

    print(f"ok    {len(material)} file(s) changed under {PLUGIN_DIR}, and {was} -> {version}")
    return True


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument(
        "--base",
        help="Ref to compare against (a pull request's base). Omit to check markers only.",
    )
    args = ap.parse_args()

    manifest = REPO_ROOT / MANIFEST
    if not manifest.exists():
        fail(f"{MANIFEST} not found.")
        return 2
    version = version_of(manifest.read_text(encoding="utf-8"), str(MANIFEST))
    if version is None:
        return 2

    ok = check_markers(version)
    if args.base:
        ok = check_moved(args.base, version) and ok
    else:
        print("skip  version-moved check (no --base given, so there is nothing to compare)")

    print("\nall checks passed" if ok else "\nchecks failed")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check_plugin_version.mjs
#!/usr/bin/env node
// The version has to move when the plugin does.
//
// Faithful port of check_plugin_version.py. Two invariants:
//   MARKERS — every SKILL.md's `vibeassist-skill-version` marker equals the
//             version in plugin.json (always runs).
//   MOVED   — if anything under plugins/ changed against the base branch, the
//             version in plugin.json must differ from the base's (needs --base).
//
// Exit 0 = every invariant it could check held. Exit 1 = a real disagreement.
// Exit 2 = the check could not run (bad ref, unreadable manifest).

import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath, pathToFileURL } from "node:url";

const SCRIPT = fs.realpathSync(fileURLToPath(import.meta.url));
const REPO_ROOT = path.dirname(path.dirname(SCRIPT));
const PLUGIN_DIR = "plugins/vibeassist";
const MANIFEST = PLUGIN_DIR + "/.claude-plugin/plugin.json";

// The marker each packaged SKILL.md carries. Captured so a mismatch can name
// both numbers rather than just saying "wrong".
const MARKER = /<!--\s*vibeassist-skill-version:\s*([0-9][0-9.]*)/;

// Files under the plugin that do NOT change what an installed copy does.
const IGNORED = new Set([MANIFEST]);

function print(s) {
  process.stdout.write(s + "\n");
}

function fail(msg) {
  print(`FAIL  ${msg}`);
}

function runGit(args) {
  const p = spawnSync("git", args, {
    cwd: REPO_ROOT,
    encoding: "utf-8",
  });
  const code = p.status === null ? 1 : p.status;
  return [code, (p.stdout || "") + (p.stderr || "")];
}

function versionOf(text, where) {
  let obj;
  try {
    obj = JSON.parse(text);
  } catch (e) {
    fail(`${where} is not valid JSON: ${e.message}`);
    return null;
  }
  const v = obj == null ? undefined : obj.version;
  if (typeof v !== "string" || v.trim() === "") {
    fail(`${where} has no \`version\` string.`);
    return null;
  }
  return v.trim();
}

// PurePosixPath ordering: compare path segment tuples lexicographically,
// element by element; a prefix tuple sorts before its extension.
function comparePosix(a, b) {
  const pa = a.split("/");
  const pb = b.split("/");
  const n = Math.min(pa.length, pb.length);
  for (let i = 0; i < n; i++) {
    if (pa[i] < pb[i]) return -1;
    if (pa[i] > pb[i]) return 1;
  }
  return pa.length - pb.length;
}

// Emulate sorted((REPO_ROOT/PLUGIN_DIR/"skills").glob("*/SKILL.md")).
function findSkillMarkers() {
  const skillsRoot = path.join(REPO_ROOT, PLUGIN_DIR, "skills");
  let entries;
  try {
    entries = fs.readdirSync(skillsRoot, { withFileTypes: true });
  } catch {
    return [];
  }
  const found = [];
  for (const ent of entries) {
    // pathlib glob "*/" matches any entry (including hidden) that is a
    // directory (or a symlink to one).
    let isDir = ent.isDirectory();
    if (!isDir && ent.isSymbolicLink()) {
      try {
        isDir = fs.statSync(path.join(skillsRoot, ent.name)).isDirectory();
      } catch {
        isDir = false;
      }
    }
    if (!isDir) continue;
    const skillMd = path.join(skillsRoot, ent.name, "SKILL.md");
    if (fs.existsSync(skillMd)) found.push(skillMd);
  }
  found.sort(comparePosix);
  return found;
}

function relToRoot(abs) {
  return path.relative(REPO_ROOT, abs).split(path.sep).join("/");
}

function checkMarkers(version) {
  const skills = findSkillMarkers();
  if (skills.length === 0) {
    fail(`no SKILL.md found under ${PLUGIN_DIR}/skills — has the layout moved?`);
    return false;
  }

  let ok = true;
  for (const p of skills) {
    const rel = relToRoot(p);
    const text = fs.readFileSync(p, "utf-8");
    const m = MARKER.exec(text);
    if (!m) {
      fail(
        `${rel} carries no \`vibeassist-skill-version\` marker, so an installed ` +
          `copy cannot tell the app which version it is running.`
      );
      ok = false;
      continue;
    }
    const found = m[1];
    if (found !== version) {
      fail(
        `${rel} says ${found}, but ${MANIFEST} says ${version}. ` +
          `One of them is lying to whoever installed it.`
      );
      ok = false;
    }
  }
  if (ok) {
    print(`ok    ${skills.length} SKILL.md marker(s) agree with ${MANIFEST}: ${version}`);
  }
  return ok;
}

function checkMoved(base, version) {
  let [code, out] = runGit(["rev-parse", "--verify", `${base}^{commit}`]);
  if (code !== 0) {
    fail(`cannot resolve base ref ${pyRepr(base)} — nothing to compare against.\n${out.trim()}`);
    process.exit(2);
  }

  [code, out] = runGit(["diff", "--name-only", `${base}...HEAD`, "--", PLUGIN_DIR]);
  if (code !== 0) {
    fail(`could not diff against ${base}.\n${out.trim()}`);
    process.exit(2);
  }

  const changed = out.split("\n").filter((line) => line.trim() !== "");
  const material = changed.filter((p) => !IGNORED.has(p));
  if (material.length === 0) {
    print(`ok    nothing under ${PLUGIN_DIR} changed against ${base} — no bump needed`);
    return true;
  }

  let baseManifest;
  [code, baseManifest] = runGit(["show", `${base}:${MANIFEST}`]);
  if (code !== 0) {
    // The manifest not existing on the base is a new plugin, not a stalled
    // version. Nothing to compare, nothing to complain about.
    print(`ok    ${MANIFEST} is new against ${base} — no previous version to move from`);
    return true;
  }

  const was = versionOf(baseManifest, `${MANIFEST} at ${base}`);
  if (was === null) {
    process.exit(2);
  }

  if (was === version) {
    const listing = material.map((p) => `        ${p}`).join("\n");
    fail(
      `${material.length} file(s) under ${PLUGIN_DIR} changed, but the version is still ` +
        `${version}:\n${listing}\n` +
        `        Anyone already on ${version} keeps a different ${version} and ` +
        `\`/plugin update vibeassist\` has nothing to move to.\n` +
        `        Bump \`version\` in ${MANIFEST} and the marker in every SKILL.md.`
    );
    return false;
  }

  print(`ok    ${material.length} file(s) changed under ${PLUGIN_DIR}, and ${was} -> ${version}`);
  return true;
}

// Python repr() of a str for the one place the source interpolates {base!r}.
function pyRepr(s) {
  const hasSingle = s.includes("'");
  const hasDouble = s.includes('"');
  const quote = hasSingle && !hasDouble ? '"' : "'";
  let body = "";
  for (const ch of s) {
    if (ch === "\\") body += "\\\\";
    else if (ch === quote) body += "\\" + ch;
    else if (ch === "\n") body += "\\n";
    else if (ch === "\r") body += "\\r";
    else if (ch === "\t") body += "\\t";
    else body += ch;
  }
  return quote + body + quote;
}

function parseArgs(argv) {
  // Minimal faithful handling of `--base` (the only real option). Argparse
  // error/help formatting is not byte-parity matched (prog name differs).
  let base = null;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--base") {
      if (i + 1 >= argv.length) {
        process.stderr.write("error: argument --base: expected one argument\n");
        process.exit(2);
      }
      base = argv[++i];
    } else if (a.startsWith("--base=")) {
      base = a.slice("--base=".length);
    } else if (a === "-h" || a === "--help") {
      print("usage: check_plugin_version.py [-h] [--base BASE]");
      process.exit(0);
    } else {
      process.stderr.write(`error: unrecognized arguments: ${a}\n`);
      process.exit(2);
    }
  }
  return { base };
}

function main() {
  const args = parseArgs(process.argv.slice(2));

  const manifest = path.join(REPO_ROOT, MANIFEST);
  if (!fs.existsSync(manifest)) {
    fail(`${MANIFEST} not found.`);
    return 2;
  }
  const version = versionOf(fs.readFileSync(manifest, "utf-8"), MANIFEST);
  if (version === null) {
    return 2;
  }

  let ok = checkMarkers(version);
  if (args.base) {
    ok = checkMoved(args.base, version) && ok;
  } else {
    print("skip  version-moved check (no --base given, so there is nothing to compare)");
  }

  print(ok ? "\nall checks passed" : "\nchecks failed");
  return ok ? 0 : 1;
}

if (import.meta.url === pathToFileURL(process.argv[1] || "").href) {
  process.exit(main());
}
